### PR TITLE
MON-2193: pkg/manifests: Expose retention size settings for Platform Prometheus

### PR DESCRIPTION
**Pull request by slashpai**
[MON-2216](https://issues.redhat.com/browse/MON-2216): Expose retention size settings for Platform Prometheus in the CMO configuration

Introduce a new "retentionSize" field under the prometheusK8s key
When neither retention nor retentionSize are defined, a default of 15d for time retention applies.
When "retentionSize" is defined and "retention" isn't, only size retention applies.
When "retentionSize" isn't defined and "retention" is, only time retention applies
When both time and size retention fields are defined, both apply.

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: pkg/manifests/manifests.go
Comment (PhilipGough) [inline pkg/manifests/manifests.go:1432]: I wonder should the previous check `if f.config.ClusterMonitoringConfiguration.PrometheusK8sConfig.Retention != ""` also check for zero values? 

If yes then perhaps we can rename `zeroRetentionSizeConfig` to `isUnsetRetentionConfig` or something similar

File: test/e2e/config_test.go
Comment (slashpai) [reply test/e2e/config_test.go:154]: ya mistyped

File: pkg/manifests/manifests.go
Comment (slashpai) [reply pkg/manifests/manifests.go:1432]: Trying to cover the scenario If the retention value is "", 0 or "0", it means that the value isn't defined, so `isUnsetRetentionConfig` name makes sense

**Comment by slashpai** [comment]
@PhilipGough @simonpasquier Please review when you get a chance

**Review by PhilipGough — APPROVED** [review]
/lgtm

**Comment by slashpai** [comment]
cc: @openshift/team-monitoring-qe-approvers

File: pkg/manifests/manifests_test.go
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1037]: `10Gi` isn't a valid value.

```suggestion
  retentionSize: 10GB
```

File: pkg/manifests/manifests.go
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1432]: I feel that `if !isUnsetRetentionConfig(...)` can be confusing to read with the double negation ("unless retention is not set"). I'd be more inclined to remove the intermediate function.

```suggestion
	if f.config.ClusterMonitoringConfiguration.PrometheusK8sConfig.RetentionSize != "" && f.config.ClusterMonitoringConfiguration.PrometheusK8sConfig.RetentionSize != "0" {
```

**Review by simonpasquier — COMMENTED** [review]
Eventually CMO should run more validations on the user inputs to avoid pushing bad configuration but it's something that we should address globally so I'm mostly fine with this PR.

**Comment by slashpai** [comment]
> Eventually CMO should run more validations on the user inputs to avoid pushing bad configuration but it's something that we should address globally so I'm mostly fine with this PR.

Once we add validating webhook in Prometheus operator for the fields it should be enough? Or should we add validations for user inputs on top of it?

File: pkg/manifests/manifests_test.go
Comment (slashpai) [reply pkg/manifests/manifests_test.go]: sure I updated it to 10GiB I missed it earlier

**Comment by lihongyan1** [comment]
After rentionSize unit issue is fixed: Still Two things are not as expected.
Detail see bug https://bugzilla.redhat.com/show_bug.cgi?id=2062510

1. When set retentionSize only, default retention time applies. Expected only retentionSize applies.

2. When retentionSize is set 0 or "0", which means the value isn't defined
   When rentention Time is set 0 or "0", 0 will be the valid value applied, suppose this is not correct

**Comment by slashpai** [comment]
> After rentionSize unit issue is fixed: Still Two things are not as expected. Detail see bug https://bugzilla.redhat.com/show_bug.cgi?id=2062510
> 
> 1. When set retentionSize only, default retention time applies. Expected only retentionSize applies.
> 2. When retentionSize is set 0 or "0", which means the value isn't defined
>    When rentention Time is set 0 or "0", 0 will be the valid value applied, suppose this is not correct

1st issue will be fixed after https://github.com/openshift/prometheus-operator/pull/162 merged

2nd issue havn't touched retention part on this so whatever value set for retention will be used even if it is `0`, I think we should add condition similar as retentionSize for retention too

@simonpasquier wdyt?

**Comment by simonpasquier** [comment]
> When retentionSize is set 0 or "0", which means the value isn't defined
When rentention Time is set 0 or "0", 0 will be the valid value applied, suppose this is not correct

This is a very specific edge case :) but Prometheus would handle it by defaulting to a retention period of 15 days.

https://github.com/prometheus/prometheus/blob/1635deed7637bc5b3616c66f4de9c2add5973eaa/cmd/prometheus/main.go#L477-L480

I'm not sure we should be doing something to prevent this configuration in CMO.  As @slashpai said, I believe that it is the same behavior in all previous OCP versions.

**Comment by jan--f** [comment]
/retitle MON-2193: pkg/manifests: Expose retention size settings for Platform Prometheus

**Comment by slashpai** [comment]
@lihongyan1 can you test this again now?

File: pkg/manifests/manifests.go
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1433]: Going back to this I'm not sure that we need to treat "0" in a special way.

**Review by simonpasquier — COMMENTED** [review]
`Config.applyDefaults()` would need an update too: DefaultRetentionValue should only be applied when both time and size retentions are empty.

/hold

File: pkg/manifests/manifests_test.go
Comment (slashpai) [reply pkg/manifests/manifests_test.go]: added

**Comment by slashpai** [comment]
@simonpasquier I updated code and added new unit test as well

@lihongyan1 could you please test again? :)

File: pkg/manifests/manifests.go
Comment (slashpai) [reply pkg/manifests/manifests.go]: I removed the check for 0

**Comment by slashpai** [comment]
@simonpasquier Can you review again? :)

**Comment by lihongyan1** [comment]
Filed a bug https://bugzilla.redhat.com/show_bug.cgi?id=2073972

**Comment by slashpai** [comment]
> Filed a bug https://bugzilla.redhat.com/show_bug.cgi?id=2073972

We had added openapi validation at CRD level to retentionSize [pkg/apis/monitoring/v1: Add validations for size settings at CRD level](https://github.com/prometheus-operator/prometheus-operator/pull/4661) in prometheus-operator

That change is synced to cmo as well as part of [kube-prometheus update](https://github.com/openshift/cluster-monitoring-operator/pull/1615/files).  So RetentionSize now is validated by the API server. If the value is invalid CMO throws an error when applying the Prometheus spec, since its rejected prometheus-operator never even gets to see the value hence no logging is done there.

Once https://github.com/prometheus-operator/prometheus-operator/pull/4684 is merged you will see same effect for retention as well

cc: @simonpasquier

File: pkg/manifests/manifests_test.go
Comment (JoaoBraveCoding) [inline pkg/manifests/manifests_test.go:1604]: nit, wdyt about moving the `err != nil`  from lines `1620-1623` to here? Makes the test easier to read no?

**Review by JoaoBraveCoding — COMMENTED** [review]
Apart from small comment it looks good 👍

File: pkg/manifests/manifests_test.go
Comment (slashpai) [reply pkg/manifests/manifests_test.go:1604]: ya definitely :)

I have updated the code

**Comment by slashpai** [comment]
> Apart from small comment it looks good 👍

Thank you :), I have updated the test

**Review by sthaha — COMMENTED** [review]
/lgtm 

I think string comparison `== ""` or " != ""` should make use of Trim(). Applicable also when you set it. 
```go
RetentionSize = strings.Trim(config.RetentionSize)
```

**Comment by slashpai** [comment]
> /lgtm
> 
> I think string comparison `== ""` or " != ""` should make use of Trim(). Applicable also when you set it.
> 
> ```go
> RetentionSize = strings.Trim(config.RetentionSize)
> ```

Good suggestion, I will update :)

**Comment by slashpai** [comment]
> > /lgtm
> > I think string comparison `== ""` or " != ""` should make use of Trim(). Applicable also when you set it.
> > ```go
> > RetentionSize = strings.Trim(config.RetentionSize)
> > ```
> 
> Good suggestion, I will update :)

Also we have other fields as well which will be good to use trim, can do that in separate PR?

**Pull request merged**
